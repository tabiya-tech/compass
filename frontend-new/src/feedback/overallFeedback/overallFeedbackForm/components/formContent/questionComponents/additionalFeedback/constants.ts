//  this is in a file of its own instead of in the AdditionalFeedbackQuestion.tsx to avoid circular dependencies
export const ADDITIONAL_FEEDBACK_QUESTION_ID = "additional_feedback";