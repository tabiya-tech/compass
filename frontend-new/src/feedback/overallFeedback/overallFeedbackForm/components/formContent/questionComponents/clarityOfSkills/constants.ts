//  this is in a file of its own instead of in the ClarityOfSkillsQuestion.tsx to avoid circular dependencies
export const CLARITY_OF_SKILLS_QUESTION_ID = "clarity_of_skills";