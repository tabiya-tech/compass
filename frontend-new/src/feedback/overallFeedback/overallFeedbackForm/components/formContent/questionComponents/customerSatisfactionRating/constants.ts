//  this is in a file of its own instead of in the CustomerSatisfaction.tsx to avoid circular dependencies
// since the question id is used in multiple places, including perhaps in the userPreferencesStateService
// which children of this component might import...
export const CUSTOMER_SATISFACTION_QUESTION_KEY = "satisfaction_with_compass";