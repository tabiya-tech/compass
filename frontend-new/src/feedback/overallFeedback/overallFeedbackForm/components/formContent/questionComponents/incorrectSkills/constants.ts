//  this is in a file of its own instead of in the IncorrectSkillsQuestion.tsx to avoid circular dependencies
export const INCORRECT_SKILLS_QUESTION_ID = "incorrect_skills";