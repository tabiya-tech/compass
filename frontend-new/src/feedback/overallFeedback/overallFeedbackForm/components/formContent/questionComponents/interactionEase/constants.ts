//  this is in a file of its own instead of in the InteractionEaseQuestion.tsx to avoid circular dependencies
export const INTERACTION_EASE_QUESTION_ID = "interaction_ease";