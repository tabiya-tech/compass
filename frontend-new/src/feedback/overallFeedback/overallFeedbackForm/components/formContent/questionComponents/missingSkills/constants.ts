//  this is in a file of its own instead of in the MissingSkillsQuestion.tsx to avoid circular dependencies
export const MISSING_SKILLS_QUESTION_ID = "missing_skills";