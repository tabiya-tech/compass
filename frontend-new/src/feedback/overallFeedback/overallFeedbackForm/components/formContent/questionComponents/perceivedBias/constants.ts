//  this is in a file of its own instead of in the PerceivedBiasQuestion.tsx to avoid circular dependencies
export const PERCEIVED_BIAS_QUESTION_ID = "perceived_bias";