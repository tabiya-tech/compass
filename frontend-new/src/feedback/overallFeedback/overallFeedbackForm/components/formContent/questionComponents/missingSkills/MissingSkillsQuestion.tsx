import React from "react";
import { Box } from "@mui/material";
import { useFeedback } from "src/feedback/overallFeedback/feedbackContext/FeedbackContext";
import QuestionRenderer
  from "src/feedback/overallFeedback/overallFeedbackForm/components/questionsRenderer/QuestionRenderer";
import { FeedbackItem } from "src/feedback/overallFeedback/overallFeedbackService/OverallFeedback.service.types";
import { FeedbackError } from "src/error/commonErrors";
import { MISSING_SKILLS_QUESTION_ID } from "./constants";

const uniqueId = "95ccb7e6-350e-4fd1-81d0-96c5b6094ab4";

export const DATA_TEST_ID = {
  MISSING_SKILLS: `missing-skills-${uniqueId}`,
};

interface MissingSkillsQuestionProps {
  feedbackItems: FeedbackItem[];
  onChange: (data: FeedbackItem) => void;
}

const MissingSkillsQuestion: React.FC<MissingSkillsQuestionProps> = ({ feedbackItems, onChange }) => {
  const { questionsConfig,error } = useFeedback();

  if (!questionsConfig) {
    console.error(new FeedbackError("Questions configuration is not available", error));
    return null;
  }

  const question = questionsConfig[MISSING_SKILLS_QUESTION_ID];

  if (!question) {
    console.error(new FeedbackError(`Questions configuration is not available for question: ${MISSING_SKILLS_QUESTION_ID}`, error));
    return null;
  }

  return (
    <Box data-testid={DATA_TEST_ID.MISSING_SKILLS}>
      <QuestionRenderer question={question} feedbackItems={feedbackItems} onChange={onChange} />
    </Box>
  );
};

export default MissingSkillsQuestion; 