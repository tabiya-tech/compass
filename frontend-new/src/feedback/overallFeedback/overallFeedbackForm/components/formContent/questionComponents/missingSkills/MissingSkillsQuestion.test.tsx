// mute the console
import "src/_test_utilities/consoleMock";
import React from "react";
import { render, screen } from "src/_test_utilities/test-utils";
import MissingSkillsQuestion, {
  DATA_TEST_ID,
} from "src/feedback/overallFeedback/overallFeedbackForm/components/formContent/questionComponents/missingSkills/MissingSkillsQuestion";
import { FeedbackProvider } from "src/feedback/overallFeedback/feedbackContext/FeedbackContext";
import { mockQuestionsConfig } from "src/feedback/overallFeedback/overallFeedbackForm/overallFeedbackForm.test.utils";
import { MISSING_SKILLS_QUESTION_ID } from "./constants";

jest.mock("src/feedback/overallFeedback/overallFeedbackForm/components/questionsRenderer/QuestionRenderer", () => {
  const actual = jest.requireActual("src/feedback/overallFeedback/overallFeedbackForm/components/questionsRenderer/QuestionRenderer");
  return {
    ...actual,
    __esModule: true,
    default: jest.fn().mockImplementation(() => <div data-testid={actual.DATA_TEST_ID.QUESTION_RENDERER}></div>),
  };
})

describe("MissingSkillsQuestion", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const mockOnChange = jest.fn();

  // Spy on the useFeedback hook to return our mockQuestionsConfig
  jest.spyOn(require("src/feedback/overallFeedback/feedbackContext/FeedbackContext"), "useFeedback").mockReturnValue({
    questionsConfig: mockQuestionsConfig,
    setQuestionsConfig: jest.fn(),
    isLoading: false,
    error: null,
    answers: [],
    handleAnswerChange: jest.fn(),
    clearAnswers: jest.fn(),
  });

  test("should render correctly with empty feedbackItems", () => {
    // GIVEN a questions config and no feedback items
    function Wrapper({ children }: React.PropsWithChildren<{}>) {
      return <FeedbackProvider>{children}</FeedbackProvider>;
    }
    // WHEN the component is rendered
    const { container } = render(
      <MissingSkillsQuestion feedbackItems={[]} onChange={mockOnChange} />, { wrapper: Wrapper }
    );
    // THEN the question should be in the document
    expect(screen.getByTestId(DATA_TEST_ID.MISSING_SKILLS)).toBeInTheDocument();
    // AND it should match the snapshot
    expect(container).toMatchSnapshot();
  });

  test("should render correctly with existing feedbackItems", () => {
    // GIVEN a questions config and a feedback item
    const feedbackItems = [
      {
        question_id: MISSING_SKILLS_QUESTION_ID,
        simplified_answer: {
          rating_boolean: true,
          comment: "Some skills were missing.",
        },
      },
    ];
    function Wrapper({ children }: React.PropsWithChildren<{}>) {
      return <FeedbackProvider>{children}</FeedbackProvider>;
    }
    // WHEN the component is rendered
    const { container } = render(
      <MissingSkillsQuestion feedbackItems={feedbackItems} onChange={mockOnChange} />, { wrapper: Wrapper }
    );
    // THEN the question should be in the document
    expect(screen.getByTestId(DATA_TEST_ID.MISSING_SKILLS)).toBeInTheDocument();
    // AND it should match the snapshot
    expect(container).toMatchSnapshot();
  });

  test("should return null when questionsConfig is not available", () => {
    // GIVEN the useFeedback hook returns null for questionsConfig
    jest.spyOn(require("src/feedback/overallFeedback/feedbackContext/FeedbackContext"), "useFeedback").mockReturnValue({
      questionsConfig: null,
      setQuestionsConfig: jest.fn(),
      isLoading: false,
      error: null,
      answers: [],
      handleAnswerChange: jest.fn(),
      clearAnswers: jest.fn(),
    });

    function Wrapper({ children }: React.PropsWithChildren<{}>) {
      return <FeedbackProvider>{children}</FeedbackProvider>;
    }

    // WHEN the component is rendered
    const { container } = render(
      <MissingSkillsQuestion feedbackItems={[]} onChange={mockOnChange} />, { wrapper: Wrapper }
    );

    // THEN expect the container to be empty
    expect(container).toBeEmptyDOMElement();
    // AND expect an error to be logged
    expect(console.error).toHaveBeenCalledWith(expect.any(Error));
  });

  test("should return null when question ID is not in questionsConfig", () => {
    // GIVEN the useFeedback hook returns a config without the required question ID
    const configWithoutQuestion = { ...mockQuestionsConfig };
    delete configWithoutQuestion[MISSING_SKILLS_QUESTION_ID];

    jest.spyOn(require("src/feedback/overallFeedback/feedbackContext/FeedbackContext"), "useFeedback").mockReturnValue({
      questionsConfig: configWithoutQuestion,
      setQuestionsConfig: jest.fn(),
      isLoading: false,
      error: null,
      answers: [],
      handleAnswerChange: jest.fn(),
      clearAnswers: jest.fn(),
    });

    function Wrapper({ children }: React.PropsWithChildren<{}>) {
      return <FeedbackProvider>{children}</FeedbackProvider>;
    }

    // WHEN the component is rendered
    const { container } = render(
      <MissingSkillsQuestion feedbackItems={[]} onChange={mockOnChange} />, { wrapper: Wrapper }
    );

    // THEN expect the container to be empty
    expect(container).toBeEmptyDOMElement();
    // AND expect an error to be logged
    expect(console.error).toHaveBeenCalledWith(expect.any(Error));
  });
}); 