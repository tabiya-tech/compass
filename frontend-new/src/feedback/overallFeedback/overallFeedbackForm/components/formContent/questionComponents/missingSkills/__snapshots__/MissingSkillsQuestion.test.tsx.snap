// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MissingSkillsQuestion should render correctly with empty feedbackItems 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
    data-testid="missing-skills-95ccb7e6-350e-4fd1-81d0-96c5b6094ab4"
  >
    <div
      data-testid="question-renderer-3bf900a2-f9fa-4b28-8c20-8bc21570635c"
    />
  </div>
</div>
`;

exports[`MissingSkillsQuestion should render correctly with existing feedbackItems 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
    data-testid="missing-skills-95ccb7e6-350e-4fd1-81d0-96c5b6094ab4"
  >
    <div
      data-testid="question-renderer-3bf900a2-f9fa-4b28-8c20-8bc21570635c"
    />
  </div>
</div>
`;
